File: analyze_results.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Metrics represents the parsed metrics from a hey output file
type Metrics struct {
	TotalRequests        int                       `json:"total_requests"`
	TotalTime            float64                   `json:"total_time"`
	RequestsPerSec       float64                   `json:"requests_per_sec"`
	AvgResponseTime      float64                   `json:"avg_response_time"`
	Slowest              float64                   `json:"slowest"`
	Fastest              float64                   `json:"fastest"`
	LatencyDistribution  map[string]float64        `json:"latency_distribution"`
	StatusCodes          map[string]int            `json:"status_codes"`
	Errors               map[string]int            `json:"errors"`
}

// BenchmarkData holds all benchmark results organized by gateway, endpoint, and concurrency
type BenchmarkData map[string]map[string]map[int]*Metrics

// parseHeyOutput parses a hey output file and extracts metrics
func parseHeyOutput(filepath string) (*Metrics, error) {
	content, err := os.ReadFile(filepath)
	if err != nil {
		return nil, err
	}

	text := string(content)
	metrics := &Metrics{
		LatencyDistribution: make(map[string]float64),
		StatusCodes:         make(map[string]int),
		Errors:              make(map[string]int),
	}

	// Extract total requests and time
	re := regexp.MustCompile(`(\d+) requests in ([\d.]+)s`)
	if match := re.FindStringSubmatch(text); match != nil {
		metrics.TotalRequests, _ = strconv.Atoi(match[1])
		metrics.TotalTime, _ = strconv.ParseFloat(match[2], 64)
	}

	// Extract requests/sec
	re = regexp.MustCompile(`Requests/sec:\s+([\d.]+)`)
	if match := re.FindStringSubmatch(text); match != nil {
		metrics.RequestsPerSec, _ = strconv.ParseFloat(match[1], 64)
	}

	// Extract average response time
	re = regexp.MustCompile(`Average:\s+([\d.]+) secs`)
	if match := re.FindStringSubmatch(text); match != nil {
		metrics.AvgResponseTime, _ = strconv.ParseFloat(match[1], 64)
	}

	// Extract slowest response
	re = regexp.MustCompile(`Slowest:\s+([\d.]+) secs`)
	if match := re.FindStringSubmatch(text); match != nil {
		metrics.Slowest, _ = strconv.ParseFloat(match[1], 64)
	}

	// Extract fastest response
	re = regexp.MustCompile(`Fastest:\s+([\d.]+) secs`)
	if match := re.FindStringSubmatch(text); match != nil {
		metrics.Fastest, _ = strconv.ParseFloat(match[1], 64)
	}

	// Extract latency distribution
	re = regexp.MustCompile(`\s+([\d.]+)%\s+in\s+([\d.]+)\s+secs`)
	for _, match := range re.FindAllStringSubmatch(text, -1) {
		percentile := match[1]
		time, _ := strconv.ParseFloat(match[2], 64)
		metrics.LatencyDistribution["p"+percentile] = time
	}

	// Extract status code distribution
	re = regexp.MustCompile(`\[(\d+)\]\s+(\d+) responses`)
	statusSection := extractSection(text, "Status code distribution:")
	for _, match := range re.FindAllStringSubmatch(statusSection, -1) {
		code := match[1]
		count, _ := strconv.Atoi(match[2])
		metrics.StatusCodes[code] = count
	}

	// Extract error distribution
	re = regexp.MustCompile(`\[(\d+)\]\s+(.+)`)
	errorSection := extractSection(text, "Error distribution:")
	for _, match := range re.FindAllStringSubmatch(errorSection, -1) {
		count, _ := strconv.Atoi(match[1])
		errorMsg := strings.TrimSpace(match[2])
		metrics.Errors[errorMsg] = count
	}

	return metrics, nil
}

// extractSection extracts a section of text after a header
func extractSection(text, header string) string {
	lines := strings.Split(text, "\n")
	inSection := false
	var section []string

	for _, line := range lines {
		if strings.Contains(line, header) {
			inSection = true
			continue
		}
		if inSection {
			if strings.TrimSpace(line) == "" || (len(line) > 0 && line[0] != ' ' && line[0] != '\t' && line[0] != '[') {
				break
			}
			section = append(section, line)
		}
	}

	return strings.Join(section, "\n")
}

// loadResults loads all result files from the results directory
func loadResults(resultsDir string) (BenchmarkData, error) {
	data := make(BenchmarkData)
	pattern := regexp.MustCompile(`test_results_(.+?)_(test|auth)_(\d+)\.txt`)

	files, err := filepath.Glob(filepath.Join(resultsDir, "test_results_*.txt"))
	if err != nil {
		return nil, err
	}

	if len(files) == 0 {
		return nil, fmt.Errorf("no result files found in '%s'", resultsDir)
	}

	for _, file := range files {
		basename := filepath.Base(file)
		match := pattern.FindStringSubmatch(basename)
		if match == nil {
			continue
		}

		gateway := match[1]
		endpoint := match[2]
		concurrency, _ := strconv.Atoi(match[3])

		metrics, err := parseHeyOutput(file)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing %s: %v\n", basename, err)
			continue
		}

		if data[gateway] == nil {
			data[gateway] = make(map[string]map[int]*Metrics)
		}
		if data[gateway][endpoint] == nil {
			data[gateway][endpoint] = make(map[int]*Metrics)
		}
		data[gateway][endpoint][concurrency] = metrics
	}

	return data, nil
}

// generateMarkdownReport generates a comprehensive markdown report
func generateMarkdownReport(data BenchmarkData) string {
	if len(data) == 0 {
		return "No benchmark data available.\n"
	}

	var report strings.Builder

	report.WriteString("# API Gateway Benchmark Results\n\n")
	report.WriteString("Total Requests: 100,000 per test\n\n")
	report.WriteString("Endpoints Tested: /test (unauthenticated), /auth (JWT-protected)\n\n")
	report.WriteString("Concurrency Levels: 50, 100, 250, 500\n\n")

	gateways := getSortedGateways(data)
	endpoints := []string{"test", "auth"}
	concurrencyLevels := []int{50, 100, 250, 500}

	// Summary table for each endpoint
	for _, endpoint := range endpoints {
		report.WriteString(fmt.Sprintf("## Results for /%s\n\n", endpoint))

		// Requests/sec comparison
		report.WriteString("### Throughput (Requests/sec)\n\n")
		report.WriteString("| Gateway | C=50 | C=100 | C=250 | C=500 |\n")
		report.WriteString("|---------|------|-------|-------|-------|\n")

		for _, gateway := range gateways {
			row := []string{gateway}
			for _, c := range concurrencyLevels {
				if metrics, ok := getMetrics(data, gateway, endpoint, c); ok {
					row = append(row, fmt.Sprintf("%.2f", metrics.RequestsPerSec))
				} else {
					row = append(row, "N/A")
				}
			}
			report.WriteString(fmt.Sprintf("| %s |\n", strings.Join(row, " | ")))
		}
		report.WriteString("\n")

		// Average response time comparison
		report.WriteString("### Average Response Time (seconds)\n\n")
		report.WriteString("| Gateway | C=50 | C=100 | C=250 | C=500 |\n")
		report.WriteString("|---------|------|-------|-------|-------|\n")

		for _, gateway := range gateways {
			row := []string{gateway}
			for _, c := range concurrencyLevels {
				if metrics, ok := getMetrics(data, gateway, endpoint, c); ok {
					row = append(row, fmt.Sprintf("%.4f", metrics.AvgResponseTime))
				} else {
					row = append(row, "N/A")
				}
			}
			report.WriteString(fmt.Sprintf("| %s |\n", strings.Join(row, " | ")))
		}
		report.WriteString("\n")

		// Latency percentiles for highest concurrency
		report.WriteString("### Latency Distribution at C=500 (seconds)\n\n")
		report.WriteString("| Gateway | 10% | 50% | 75% | 90% | 95% | 99% |\n")
		report.WriteString("|---------|-----|-----|-----|-----|-----|-----|\n")

		for _, gateway := range gateways {
			row := []string{gateway}
			if metrics, ok := getMetrics(data, gateway, endpoint, 500); ok {
				for _, p := range []string{"10", "50", "75", "90", "95", "99"} {
					key := "p" + p
					if val, exists := metrics.LatencyDistribution[key]; exists {
						row = append(row, fmt.Sprintf("%.4f", val))
					} else {
						row = append(row, "N/A")
					}
				}
			} else {
				for i := 0; i < 6; i++ {
					row = append(row, "N/A")
				}
			}
			report.WriteString(fmt.Sprintf("| %s |\n", strings.Join(row, " | ")))
		}
		report.WriteString("\n")
	}

	// Winner analysis
	report.WriteString("## Performance Summary\n\n")

	for _, endpoint := range endpoints {
		report.WriteString(fmt.Sprintf("### /%s Endpoint\n\n", endpoint))

		// Find best throughput at each concurrency
		for _, c := range concurrencyLevels {
			bestGateway := ""
			bestRPS := 0.0

			for _, gateway := range gateways {
				if metrics, ok := getMetrics(data, gateway, endpoint, c); ok {
					if metrics.RequestsPerSec > bestRPS {
						bestRPS = metrics.RequestsPerSec
						bestGateway = gateway
					}
				}
			}

			if bestGateway != "" {
				report.WriteString(fmt.Sprintf("**Concurrency %d**: %s - %.2f req/s\n\n", c, bestGateway, bestRPS))
			}
		}
	}

	// Error analysis
	report.WriteString("## Error Analysis\n\n")
	hasErrors := false

	for _, gateway := range gateways {
		for _, endpoint := range endpoints {
			for _, c := range concurrencyLevels {
				if metrics, ok := getMetrics(data, gateway, endpoint, c); ok {
					// Check for non-200 status codes
					errorCodes := make(map[string]int)
					for code, count := range metrics.StatusCodes {
						if code != "200" {
							errorCodes[code] = count
						}
					}

					if len(metrics.Errors) > 0 || len(errorCodes) > 0 {
						hasErrors = true
						report.WriteString(fmt.Sprintf("**%s** (/%s, C=%d):\n", gateway, endpoint, c))
						if len(errorCodes) > 0 {
							report.WriteString(fmt.Sprintf("  - Status codes: %v\n", errorCodes))
						}
						if len(metrics.Errors) > 0 {
							report.WriteString(fmt.Sprintf("  - Errors: %v\n", metrics.Errors))
						}
						report.WriteString("\n")
					}
				}
			}
		}
	}

	if !hasErrors {
		report.WriteString("No errors detected in any test runs.\n\n")
	}

	return report.String()
}

// getSortedGateways returns a sorted list of gateway names
func getSortedGateways(data BenchmarkData) []string {
	gateways := make([]string, 0, len(data))
	for gateway := range data {
		gateways = append(gateways, gateway)
	}
	sort.Strings(gateways)
	return gateways
}

// getMetrics retrieves metrics for a specific gateway, endpoint, and concurrency level
func getMetrics(data BenchmarkData, gateway, endpoint string, concurrency int) (*Metrics, bool) {
	if data[gateway] == nil {
		return nil, false
	}
	if data[gateway][endpoint] == nil {
		return nil, false
	}
	metrics, ok := data[gateway][endpoint][concurrency]
	return metrics, ok
}

func main() {
	resultsDir := "results"

	// Load results
	data, err := loadResults(resultsDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading results: %v\n", err)
		os.Exit(1)
	}

	// Generate markdown report
	markdown := generateMarkdownReport(data)
	outputFile := filepath.Join(resultsDir, "BENCHMARK_REPORT.md")
	if err := os.WriteFile(outputFile, []byte(markdown), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing markdown report: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Markdown report generated: %s\n", outputFile)
	fmt.Println("\n" + strings.Repeat("=", 80) + "\n")
	fmt.Print(markdown)

	// Generate JSON report
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling JSON: %v\n", err)
		os.Exit(1)
	}
	jsonFile := filepath.Join(resultsDir, "benchmark_data.json")
	if err := os.WriteFile(jsonFile, jsonData, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing JSON report: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("JSON report generated: %s\n", jsonFile)
}

File: analyze_results.py
#!/usr/bin/env python3
"""
Benchmark Results Analyzer for API Gateway Performance Tests

Parses 'hey' output and generates comparison reports.
"""

import os
import re
import json
from pathlib import Path
from collections import defaultdict
import sys


class BenchmarkAnalyzer:
    def __init__(self, results_dir="results"):
        self.results_dir = Path(results_dir)
        self.data = defaultdict(lambda: defaultdict(dict))

    def parse_hey_output(self, filepath):
        """Parse hey output file and extract metrics."""
        with open(filepath, 'r') as f:
            content = f.read()

        metrics = {}

        # Extract total requests
        match = re.search(r'(\d+) requests in ([\d.]+)s', content)
        if match:
            metrics['total_requests'] = int(match.group(1))
            metrics['total_time'] = float(match.group(2))

        # Extract requests/sec
        match = re.search(r'Requests/sec:\s+([\d.]+)', content)
        if match:
            metrics['requests_per_sec'] = float(match.group(1))

        # Extract average response time
        match = re.search(r'Average:\s+([\d.]+) secs', content)
        if match:
            metrics['avg_response_time'] = float(match.group(1))

        # Extract slowest response
        match = re.search(r'Slowest:\s+([\d.]+) secs', content)
        if match:
            metrics['slowest'] = float(match.group(1))

        # Extract fastest response
        match = re.search(r'Fastest:\s+([\d.]+) secs', content)
        if match:
            metrics['fastest'] = float(match.group(1))

        # Extract latency distribution
        latency_dist = {}
        for line in content.split('\n'):
            match = re.match(r'\s+([\d.]+)%\s+in\s+([\d.]+)\s+secs', line)
            if match:
                percentile = match.group(1)
                time = float(match.group(2))
                latency_dist[f'p{percentile}'] = time
        metrics['latency_distribution'] = latency_dist

        # Extract status code distribution
        status_codes = {}
        match = re.search(r'Status code distribution:\s*\n((?:\s+\[.*?\]\s+\d+\s+responses\s*\n?)*)', content)
        if match:
            for line in match.group(1).split('\n'):
                code_match = re.search(r'\[(\d+)\]\s+(\d+) responses', line)
                if code_match:
                    status_codes[code_match.group(1)] = int(code_match.group(2))
        metrics['status_codes'] = status_codes

        # Extract error distribution
        errors = {}
        match = re.search(r'Error distribution:\s*\n((?:\s+\[.*?\]\s+\d+\s*\n?)*)', content)
        if match:
            for line in match.group(1).split('\n'):
                error_match = re.search(r'\[(\d+)\]\s+(.+)', line)
                if error_match:
                    errors[error_match.group(2).strip()] = int(error_match.group(1))
        metrics['errors'] = errors

        return metrics

    def load_results(self):
        """Load all result files from the results directory."""
        pattern = re.compile(r'test_results_(.+?)_(test|auth)_(\d+)\.txt')

        if not self.results_dir.exists():
            print(f"Results directory '{self.results_dir}' not found!")
            return False

        files = list(self.results_dir.glob('test_results_*.txt'))
        if not files:
            print(f"No result files found in '{self.results_dir}'!")
            return False

        for filepath in files:
            match = pattern.match(filepath.name)
            if match:
                gateway = match.group(1)
                endpoint = match.group(2)
                concurrency = int(match.group(3))

                try:
                    metrics = self.parse_hey_output(filepath)
                    self.data[gateway][endpoint][concurrency] = metrics
                except Exception as e:
                    print(f"Error parsing {filepath.name}: {e}")

        return True

    def generate_markdown_report(self):
        """Generate a comprehensive markdown report."""
        if not self.data:
            return "No benchmark data available."

        report = []
        report.append("# API Gateway Benchmark Results\n")
        report.append(f"Total Requests: 100,000 per test\n")
        report.append(f"Endpoints Tested: /test (unauthenticated), /auth (JWT-protected)\n")
        report.append(f"Concurrency Levels: 50, 100, 250, 500\n\n")

        gateways = sorted(self.data.keys())
        endpoints = ['test', 'auth']
        concurrency_levels = [50, 100, 250, 500]

        # Summary table for each endpoint
        for endpoint in endpoints:
            report.append(f"## Results for /{endpoint}\n")

            # Requests/sec comparison
            report.append(f"### Throughput (Requests/sec)\n")
            report.append("| Gateway | C=50 | C=100 | C=250 | C=500 |\n")
            report.append("|---------|------|-------|-------|-------|\n")

            for gateway in gateways:
                row = [gateway]
                for c in concurrency_levels:
                    metrics = self.data[gateway].get(endpoint, {}).get(c, {})
                    rps = metrics.get('requests_per_sec', 0)
                    row.append(f"{rps:,.2f}" if rps > 0 else "N/A")
                report.append("| " + " | ".join(row) + " |\n")
            report.append("\n")

            # Average response time comparison
            report.append(f"### Average Response Time (seconds)\n")
            report.append("| Gateway | C=50 | C=100 | C=250 | C=500 |\n")
            report.append("|---------|------|-------|-------|-------|\n")

            for gateway in gateways:
                row = [gateway]
                for c in concurrency_levels:
                    metrics = self.data[gateway].get(endpoint, {}).get(c, {})
                    avg = metrics.get('avg_response_time', 0)
                    row.append(f"{avg:.4f}" if avg > 0 else "N/A")
                report.append("| " + " | ".join(row) + " |\n")
            report.append("\n")

            # Latency percentiles for highest concurrency
            report.append(f"### Latency Distribution at C=500 (seconds)\n")
            report.append("| Gateway | 10% | 50% | 75% | 90% | 95% | 99% |\n")
            report.append("|---------|-----|-----|-----|-----|-----|-----|\n")

            for gateway in gateways:
                metrics = self.data[gateway].get(endpoint, {}).get(500, {})
                latency = metrics.get('latency_distribution', {})
                row = [gateway]
                for p in ['10', '50', '75', '90', '95', '99']:
                    val = latency.get(f'p{p}', 0)
                    row.append(f"{val:.4f}" if val > 0 else "N/A")
                report.append("| " + " | ".join(row) + " |\n")
            report.append("\n")

        # Winner analysis
        report.append("## Performance Summary\n\n")

        for endpoint in endpoints:
            report.append(f"### /{endpoint} Endpoint\n\n")

            # Find best throughput at each concurrency
            for c in concurrency_levels:
                best_gateway = None
                best_rps = 0

                for gateway in gateways:
                    metrics = self.data[gateway].get(endpoint, {}).get(c, {})
                    rps = metrics.get('requests_per_sec', 0)
                    if rps > best_rps:
                        best_rps = rps
                        best_gateway = gateway

                if best_gateway:
                    report.append(f"**Concurrency {c}**: {best_gateway} - {best_rps:,.2f} req/s\n\n")

        # Error analysis
        report.append("## Error Analysis\n\n")
        has_errors = False

        for gateway in gateways:
            for endpoint in endpoints:
                for c in concurrency_levels:
                    metrics = self.data[gateway].get(endpoint, {}).get(c, {})
                    errors = metrics.get('errors', {})
                    status_codes = metrics.get('status_codes', {})

                    # Check for non-200 status codes
                    error_codes = {k: v for k, v in status_codes.items() if k != '200'}

                    if errors or error_codes:
                        has_errors = True
                        report.append(f"**{gateway}** (/{endpoint}, C={c}):\n")
                        if error_codes:
                            report.append(f"  - Status codes: {error_codes}\n")
                        if errors:
                            report.append(f"  - Errors: {errors}\n")
                        report.append("\n")

        if not has_errors:
            report.append("No errors detected in any test runs.\n\n")

        return "".join(report)

    def generate_json_report(self):
        """Generate a JSON report of all data."""
        return json.dumps(dict(self.data), indent=2)


def main():
    analyzer = BenchmarkAnalyzer()

    if not analyzer.load_results():
        sys.exit(1)

    # Generate markdown report
    markdown = analyzer.generate_markdown_report()
    output_file = Path("results/BENCHMARK_REPORT.md")
    output_file.write_text(markdown)
    print(f"Markdown report generated: {output_file}")
    print("\n" + "="*80 + "\n")
    print(markdown)

    # Generate JSON report
    json_report = analyzer.generate_json_report()
    json_file = Path("results/benchmark_data.json")
    json_file.write_text(json_report)
    print(f"JSON report generated: {json_file}")


if __name__ == "__main__":
    main()
